router/proxy/reverseproxy.go

```diff
@@ -128,22 +128,32 @@ func (p *ReverseProxy) serveUpgrade(rw http.ResponseWriter, req *http.Request) {
 	}
 	defer uconn.Close()
 
-	p.writeResponse(rw, res)
-	res.Body.Close()
+	dconn, bufrw, err := rw.(http.Hijacker).Hijack()
```

Comment:
I think the `prepareResponseHeaders` call and the status code check can happen moved before call to hijack. That way, if `res.StatusCode != 101`, it can use `rw` to send back the response. And I believe it's OK to always call `prepareResponseHeaders`.